projects/cloud_ml/scripts/scoring.py

```diff
@@ -0,0 +1,30 @@
+import os
+import fv3viz
+import xarray as xr
+import numpy as np
+from vcm.calc.calc import weighted_average
+from vcm.calc import metrics
+from vcm import mass_integrate
+
+def find_common_times(ds1, ds2):
+    return list(set(ds1.time.values).intersection(set(ds2.time.values)))
+    
+def global_average(x):
+    return x.mean(['x', 'y','tile'])
+
+def global_sum(x):
+    return x.sum(['x', 'y','tile'])
+
+def global_variance(x):
+    ## returns the errors relative its global mean
+    return x.var(['x', 'y','tile'])
+
+def r2_score(x, y, mean = global_average):
+    var  = mean((x - mean(x))**2)
+    mse  = metrics.mean_squared_error(x, y, mean = mean)
+    return 1 - mse/var 
+
+def r2_integrated_skill(x, y, delp, z_dim = 'pressure',mean = global_average):
+    mean_integrated_variance  = mean(mass_integrate((x - mean(x))**2 , delp, z_dim))
```

Comment:
Pasting my comment from my comment in the [google doc](https://docs.google.com/document/d/1MmlolS3z_Hg_NSrR4UM2xf9A9JoLiRlc81cLgRn_Vv4/edit) here so that it's more visible:

> While I don't think it's wrong to compute it this way, it's not very intuitive to mass integrate a variance or squared error. More typically, we have either: a) mass-integrated the relevant quantities and computed the statistics on the 2D fields or b) computed the metric at each 3D level as you've done to produce the time-pressure plots, and then computed a mass-weighted average of the metric over the vertical dimension. I would probably vote for a) since it's the simplest.

What I mean by a) was that you should _first_ mass integrate the fields `x` and `y` to 2D before any squared errors are computed. Then compute the mean variance and errors on those 2D fields.

In other words, the `r2_integrated_skill` metric should be the same as the `r2_score` if the `r2_score` is computed on the mass-integrated fields (in fact, you can just call `r2_score` on the mass integrated fields inside of `r2_integrated_skill`). As currently written it's not the same due to the non-linear squaring being done at done in a different order, and also it's not as intuitive to mass integrate a squared error or variance.

Reply:
I'm not sure I agree with this. Perhaps we should devote some time tomorrow
to write this out. There is value to non mass-weighted MSEs especially for
this problem. The mass may be small aloft, but clouds there can have a
large radiative impact.

For now, maybe it's simpler to forget that the "z" dimension and all its
complexity exists...especially since Chris' first suggestion was to treat
this as a grid point local problem and "z" is just another dimension that
we take samples over.

There's been plenty of back and forth on this. I think a pair
programming/whiteboard session will be an efficient way to put this to rest.
[…](https://github.com/ai2cm/fv3net/pull/1906#)